### PR TITLE
Componentizes power cells

### PR DESCRIPTION
**Pull request by Zonespace27**
# About the pull request
Creates a new, flexible component for power cells. Eventually, anything powered will use a cell component instead of holding a ref to a `/obj/item/cell` or having a `charge` variable.

Converts NVGs over to the new system as an example, the rest will come in time.

Minor rework of how `emp_act()` works

# Explain why it's good for the game
Cells are one of the most obvious things to be componentized, and componentizing them allows cells to be added to new objects in a very easy manner.

# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

Works

</details>

### CODE REVIEW

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

File: code/datums/components/cell.dm [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+		to_chat(user, SPAN_WARNING("[parent] doesn't need more power."))
+		return
+
+	while(charge < max_charge)
+		if(SEND_SIGNAL(parent, COMSIG_CELL_TRY_RECHARGING, user) & COMPONENT_CELL_NO_RECHARGE)
+			return
+
+		if(power_cell.charge <= 0)
+			to_chat(user, SPAN_WARNING("[power_cell] is completely dry."))
+			return
```
Comment (fira) [inline code/datums/components/cell.dm:145]: For consistency i'd advise checking the charge after do_after (or both) to make sure it's valid at that point in time

**Review by fira — COMMENTED** [review]
Looks good to me codewise, will probably test it later

File: code/datums/components/cell.dm (hunk 1 above)
Comment (Zonespace27) [reply code/datums/components/cell.dm:145]: done

**Comment by fira** [comment]
Tested it, looks good to me, might just need a way to modify the examine text behavior in the future to fit objects depending on where we stick it

**Review by fira — APPROVED** [review]

**Review by harryob — APPROVED** [review]

**Pull request merged**